DEV: Redesign chat mentions

AndrewPrigorshnev opened the pull request:
At the moment, when someone is mentioning a group, or using `here` or `all` mention, we create a `chat_mention` record _per user_. What we want instead is to have special kinds of mentions, so we can create only one `chat_mention` record in such cases. This PR implements that.

Note, that such mentions will still have N related notifications, one notification per a user. We don't expect we'll have performance problems on the notifications side, but if at some point we do, we should be able to solve them on the side of notifications (notifications are handled in jobs, also some little delays with the notifications are acceptable, so we can make sure notifications are properly queued, and that processing of every notification is fast enough to make delays small enough).

The preparation work for this PR was done in fbd24fa6, where we make it possible for one mention to have several related notifications.

A pretty tricky part of this PR is schema and data migration, I've explained related details inline on the migration files.

jjaffeux commented on plugins/chat/app/models/chat/message.rb:
```diff
@@ -314,6 +318,38 @@ def message_too_long?
       self.last_editor_id ||= self.user_id
     end
+
+    def create_or_delete_all_mention
+      if !parsed_mentions.has_global_mention && all_mention.present?
+        all_mention.destroy!
+        association(:all_mention).reload
+      elsif parsed_mentions.has_global_mention && all_mention.blank?
+        build_all_mention.save!
+      end
+    end
+
+    def create_or_delete_here_mention
+      if !parsed_mentions.has_here_mention && here_mention.present?
+        here_mention.destroy!
+        association(:here_mention).reload
+      elsif parsed_mentions.has_here_mention && here_mention.blank?
+        build_here_mention.save!
+      end
+    end
+
+    def upsert_group_mentions
+      old_mentions = group_mentions.pluck(:target_id)
+      new_mentions = parsed_mentions.groups_to_mention.pluck(:id)
+      delete_mentions("Chat::GroupMention", old_mentions - new_mentions)
+      insert_mentions("Chat::GroupMention", new_mentions - old_mentions)
+    end
+
+    def upsert_user_mentions
+      old_mentions = user_mentions.pluck(:target_id)
+      new_mentions = parsed_mentions.direct_mentions.pluck(:id)
+      delete_mentions("Chat::UserMention", old_mentions - new_mentions)
+      insert_mentions("Chat::UserMention", new_mentions - old_mentions)
+    end
```
So if I follow this correctly we will in final call:  `insert_all` (and destory also maybe) multiple times if we have a message with group/user/here/all, can't we find a way to have all the inserts in one operation?

jjaffeux commented on plugins/chat/db/migrate/20231227160001_add_type_and_target_id_to_chat_mentions.rb:
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
```
I would suggest you consider all the index operations in another migration and follow this kind of pattern instead:

```ruby
# frozen_string_literal: true

class AddChatMentionsIndex < ActiveRecord::Migration[7.0]
  disable_ddl_transaction!

  def up
    remove_index :chat_mentions, name: :chat_mentions_index, algorithm: :concurrently
    add_index :chat_mentions,
              [:chat_message_id],
              algorithm: :concurrently
  end

  def down
    add_index :chat_mentions,
              [:chat_message_id, :user_id, :notification_id],
              unique: true,
              name: 'chat_mentions_index',
              algorithm: :concurrently
    remove_index :chat_mentions, column: [:chat_message_id], algorithm: :concurrently
  end
end
```

AndrewPrigorshnev replied:
This makes a lot of sense 👍.

jjaffeux commented on plugins/chat/app/jobs/regular/chat/notify_mentioned.rb:
```diff
@@ -145,13 +145,39 @@ def process_mentions(user_ids, mention_type)
+        if mention_klass == ::Chat::UserMention
+          target_id = user_id
+        elsif mention_klass == ::Chat::GroupMention
+          target_id = Group.where("LOWER(name) = ?", mention_type).first.id
```
We don't handle the cast where we don't find the group

AndrewPrigorshnev replied:
I think, the only thing we can do in that case is to write a message to logs a06deea9a810509951d26f66441202f591f56401? Is it what you had in mind? This won't handle the problem with the notification about the group mention, but at least other notifications for a chat message will be sent.

One note is that, it's highly unlikely a group won't be found here. The name of the group comes from `Notifier`, and `Notifier` makes sure that the group exists just a moment before this code runs. It's not really explicit, but `@parsed_mentions.group_mentions` in this place returns only existing groups:

https://github.com/discourse/discourse/blob/93ca13e5346aada027a7b9cadb43eba650889f20/plugins/chat/lib/chat/notifier.rb#L199-L200

That `ParsedMentions` class is a result of refactoring (we extracted that logic from Notifier), and it is still far from being ideal.

Something that I believe we can improve about the `NotifyMentioned` job is to make it work like this: the job can take the only parameter – message_id, and then it can load mentions for that message and send notifications for them. The current implementation when Notifier loads mentions, check them and then call the `NotifyMentioned` job with passing to it a hash with mentions feels too complex to me.

jjaffeux commented on plugins/chat/app/jobs/regular/chat/notify_mentioned.rb:
```diff
@@ -145,13 +145,39 @@ def process_mentions(user_ids, mention_type)
+          target_id = Group.where("LOWER(name) = ?", mention_type).first.id
+        end
+
+        mention_klass.find_by(chat_message: @chat_message, target_id: target_id)
```
not sure I like to have `@chat_message` as an implicit dependency, would probably prefer this to be an argument of `find_mention`

AndrewPrigorshnev replied:
Fixed

jjaffeux commented on plugins/chat/db/migrate/20231227160001_add_type_and_target_id_to_chat_mentions.rb:
```diff
@@ -0,0 +1,24 @@
+  def up
+    add_column :chat_mentions, :type, :string, null: true
+    add_column :chat_mentions, :target_id, :integer, null: true
+    change_column :chat_mentions, :user_id, :integer, null: true
```
Shouldn't we use `change_column_null` instead?

jjaffeux commented on plugins/chat/db/post_migrate/20231227160005_make_type_on_chat_mentions_non_nullable.rb:
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class MakeTypeOnChatMentionsNonNullable < ActiveRecord::Migration[7.0]
+  def up
+    change_column :chat_mentions, :type, :string, null: false
+  end
+
+  def down
+    change_column :chat_mentions, :type, :string, null: true
+  end
```
Should we use `change_column_null` instead?

AndrewPrigorshnev replied:
Fixed.

AndrewPrigorshnev commented:
@jjaffeux, I've rebased the branch and also fixed test failures (the most of them were because of a typo). I've also addressed your recent comment regarding `change_column_null`.

jjaffeux reviewed (approved):
Ok, didn't retry it locally this time, but it seems fine and not much has changed anyways. :rocket:

ZogStriP commented on plugins/chat/app/jobs/regular/chat/notify_mentioned.rb:
```diff
@@ -145,13 +145,43 @@ def process_mentions(user_ids, mention_type)
+          target_id = user_id
+        elsif mention_klass == ::Chat::GroupMention
+          begin
+            target_id = Group.where("LOWER(name) = ?", "#{mention_type}").first.id
```
Nitpick: if you only care about some fields about 1 record, you can use `#pick` and list the fields you want.

```suggestion
            target_id = Group.where("LOWER(name) = ?", "#{mention_type}").pick(:id)
```

AndrewPrigorshnev replied:
Sorry, merged the PR before I saw this. I've fixed it in a follow-up – https://github.com/discourse/discourse/pull/25301.
